diff -Nur ./ghemical-3.0.0.original/src/gtk_p1dview_wnd.cpp ./ghemical-3.0.0/src/gtk_p1dview_wnd.cpp
--- ./ghemical-3.0.0.original/src/gtk_p1dview_wnd.cpp	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/gtk_p1dview_wnd.cpp	2023-07-04 19:37:21.201545436 +0000
@@ -30,7 +30,8 @@
 GtkActionEntry gtk_p1dview_wnd::entries[] =
 {
 	{ "AttachDetachPlot1D", NULL, N_("Attach/Detach This View"), NULL, N_("Select whether this view is attached or in an independent window"), (GCallback) gtk_p1dview_wnd::popup_AttachDetach },
-	{ "DeleteViewPlot1D", NULL, N_("Delete View"), NULL, N_("Delete this view"), (GCallback) gtk_p1dview_wnd::popup_DeleteView }
+	{ "DeleteViewPlot1D", NULL, N_("Delete View"), NULL, N_("Delete this view"), (GCallback) gtk_p1dview_wnd::popup_DeleteView },
+	{ "SavePlot1DData", NULL, N_("Save Data"), NULL, N_("Save "), (GCallback) gtk_p1dview_wnd::popup_SaveData }
 };
 
 const char * gtk_p1dview_wnd::ui_description =
@@ -39,6 +40,8 @@
 "    <menuitem action='AttachDetachPlot1D'/>"
 "    <separator/>"
 "    <menuitem action='DeleteViewPlot1D'/>"
+"    <separator/>"
+"    <menuitem action='SavePlot1DData'/>"
 "  </popup>"
 "</ui>";
 
@@ -87,6 +90,14 @@
 	gtk_app::GetPrjX()->RemovePlottingClient(wcl);
 }
 
+void gtk_p1dview_wnd::popup_SaveData(GtkWidget *, gpointer data)
+{
+	gtk_wnd * wnd = iv_Find((GtkWidget *) data);
+	p1dview_wcl * wcl = dynamic_cast<p1dview_wcl *>(wnd->GetClient());
+	
+	gtk_app::GetPrjX()->SavePlottingClient(wcl);
+}
+
 /*################################################################################################*/
 
 // eof
diff -Nur ./ghemical-3.0.0.original/src/gtk_p1dview_wnd.h ./ghemical-3.0.0/src/gtk_p1dview_wnd.h
--- ./ghemical-3.0.0.original/src/gtk_p1dview_wnd.h	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/gtk_p1dview_wnd.h	2023-07-04 19:37:40.009544943 +0000
@@ -42,6 +42,7 @@
 	
 	static void popup_AttachDetach(GtkWidget *, gpointer);
 	static void popup_DeleteView(GtkWidget *, gpointer);
+	static void popup_SaveData(GtkWidget *, gpointer);
 };
 
 /*################################################################################################*/
diff -Nur ./ghemical-3.0.0.original/src/gtk_simple_dialogs.cpp ./ghemical-3.0.0/src/gtk_simple_dialogs.cpp
--- ./ghemical-3.0.0.original/src/gtk_simple_dialogs.cpp	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/gtk_simple_dialogs.cpp	2023-07-05 04:10:12.068737771 +0000
@@ -189,13 +189,34 @@
 
 /*################################################################################################*/
 
-const char * make_tmp_filename(gtk_project * prj)
+const char * make_tmp_filename(project * prj)
 {
 	static char tmp_filename[256];
 	prj->GetFullProjectFileName(tmp_filename, 256);
 	return tmp_filename;
 }
 
+const char * make_box_filename(project * prj)
+{
+	static char tmp_filename[256];
+	prj->GetFullBoxFileName(tmp_filename, 256);
+	return tmp_filename;
+}
+
+const char * make_traj_filename(project * prj)
+{
+	static char tmp_filename[256];
+	prj->GetFullTrajFileName(tmp_filename, 256);
+	return tmp_filename;
+}
+
+const char * make_plotdata_filename(project * prj, base_wcl * plot)
+{
+	static char tmp_filename[512];
+	prj->GetFullPlotDataFileName(tmp_filename, plot->GetTitle(), 512);
+	return tmp_filename;
+}
+
 GtkWidget * gtk_file_dialog::dialog;
 
 gtk_file_dialog::gtk_file_dialog(const char * title, const char * filter, const char * def_fn)
@@ -355,6 +376,48 @@
 
 /*################################################################################################*/
 
+const char gtk_plot_data_save_dialog::title[] = "Save Plot Data";
+project * gtk_plot_data_save_dialog::prj = NULL;
+p1dview_wcl * gtk_plot_data_save_dialog::plot = NULL;
+
+gtk_plot_data_save_dialog::gtk_plot_data_save_dialog(project * p1, p1dview_wcl * p2) :
+	gtk_file_dialog(title, PLOTFILENAME_FILTER, make_plotdata_filename(p1, p2))
+{
+	prj = p1;
+	plot = p2;
+}
+
+gtk_plot_data_save_dialog::~gtk_plot_data_save_dialog(void)
+{
+}
+
+bool gtk_plot_data_save_dialog::OkEvent(const char * filename)
+{
+	if (plot == NULL)
+	{
+		assertion_failed(__FILE__, __LINE__, "plot == NULL");
+	}
+
+	char buffer[256];
+	prj->GetProjectFileName(buffer, 256, true);
+	
+	char tmp_filename[512];
+	prj->GetFullPlotDataFileName(tmp_filename, plot->GetTitle(), 512);
+	
+	ofstream ofile;
+	ofile.open(tmp_filename, ios::out);
+	plot->SaveData(ofile);
+	ofile.close();
+
+	return true;
+}
+
+void gtk_plot_data_save_dialog::CancelEvent(void)
+{
+}
+
+/*################################################################################################*/
+
 const char gtk_file_save_graphics_dialog::title[] = "Save Graphics File - ps";
 gtk_project * gtk_file_save_graphics_dialog::prj = NULL;
 
@@ -392,7 +455,7 @@
 gtk_project * gtk_trajfile_dialog::prj = NULL;
 
 gtk_trajfile_dialog::gtk_trajfile_dialog(gtk_project * p1) :
-	gtk_file_dialog(title, NULL, NULL)
+	gtk_file_dialog(title, NULL, make_traj_filename(p1))
 {
 	prj = p1;
 }
@@ -433,7 +496,7 @@
 gtk_project * gtk_importpdb_dialog::prj = NULL;
 
 gtk_importpdb_dialog::gtk_importpdb_dialog(gtk_project * p1) :
-	gtk_file_dialog(title, NULL, NULL)
+	gtk_file_dialog(title, NULL, make_tmp_filename(p1))
 {
 	prj = p1;
 }
@@ -470,7 +533,7 @@
 gtk_project * gtk_importbox_dialog::prj = NULL;
 
 gtk_importbox_dialog::gtk_importbox_dialog(gtk_project * p1) :
-	gtk_file_dialog(title, NULL, NULL)
+	gtk_file_dialog(title, BOXFILENAME_FILTER, make_box_filename(p1))
 {
 	prj = p1;
 }
diff -Nur ./ghemical-3.0.0.original/src/gtk_simple_dialogs.h ./ghemical-3.0.0/src/gtk_simple_dialogs.h
--- ./ghemical-3.0.0.original/src/gtk_simple_dialogs.h	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/gtk_simple_dialogs.h	2023-07-04 20:46:14.669436942 +0000
@@ -74,7 +74,7 @@
 
 /*################################################################################################*/
 
-const char * make_tmp_filename(gtk_project *);
+const char * make_tmp_filename(project *);
 
 class gtk_file_dialog
 {
@@ -131,6 +131,25 @@
 	
 	bool OkEvent(const char *);	// virtual
 	void CancelEvent(void);		// virtual
+};
+
+/*################################################################################################*/
+
+class gtk_plot_data_save_dialog : public gtk_file_dialog
+{
+	private:
+	
+	static const char title[];
+	static project * prj;
+	static p1dview_wcl * plot;
+	
+	public:
+	
+	gtk_plot_data_save_dialog(project *, p1dview_wcl *);
+	~gtk_plot_data_save_dialog(void);
+	
+	bool OkEvent(const char *);	// virtual
+	void CancelEvent(void);		// virtual
 };
 
 /*################################################################################################*/
diff -Nur ./ghemical-3.0.0.original/src/p1dview_wcl.cpp ./ghemical-3.0.0/src/p1dview_wcl.cpp
--- ./ghemical-3.0.0.original/src/p1dview_wcl.cpp	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/p1dview_wcl.cpp	2023-07-05 04:45:39.104681941 +0000
@@ -75,6 +75,15 @@
 	dv.push_back(newdata);
 }
 
+void p1dview_wcl::SaveData(ofstream & file)
+{
+	file << name1 << ", " << namev << std::endl;
+	for (i32u n1 = 0;n1 < dv.size();n1++)
+	{
+		file << dv[n1].c1 << ", " << dv[n1].v << std::endl;
+	}
+}
+
 void p1dview_wcl::AddDataWithAC(double c1, double v, engine * eng)
 {
 	p1d_data newdata;
diff -Nur ./ghemical-3.0.0.original/src/p1dview_wcl.h ./ghemical-3.0.0/src/p1dview_wcl.h
--- ./ghemical-3.0.0.original/src/p1dview_wcl.h	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/p1dview_wcl.h	2023-07-04 19:38:07.725544215 +0000
@@ -61,6 +61,7 @@
 	virtual ~p1dview_wcl(void);
 	
 	void AddData(double, double);
+	void SaveData(ofstream & file);
 	
 	void AddDataWithAC(double, double, engine *);
 	void AddDataWithAC(double, double, model *, int);
diff -Nur ./ghemical-3.0.0.original/src/project.cpp ./ghemical-3.0.0/src/project.cpp
--- ./ghemical-3.0.0.original/src/project.cpp	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/project.cpp	2023-07-04 19:38:38.145543417 +0000
@@ -20,6 +20,8 @@
 
 #include "project.h"	// config.h is here -> we get ENABLE-macros here...
 
+#include "gtk_simple_dialogs.h"
+
 #include <ghemical/libghemicaldefine.h>
 
 #include <ghemical/v3d.h>
@@ -190,6 +192,24 @@
 	return ext;
 }
 
+const char * project::GetBoxFileNameExtension(void)
+{
+	static const char ext[] = "box";
+	return ext;
+}
+
+const char * project::GetTrajFileNameExtension(void)
+{
+	static const char ext[] = "traj_v17";
+	return ext;
+}
+
+const char * project::GetPlotDataFileNameExtension(void)
+{
+	static const char ext[] = "txt";
+	return ext;
+}
+
 void project::SetProjectPath(const char * path)
 {
 	if (project_path != NULL) delete[] project_path;
@@ -301,6 +321,51 @@
 	strcpy(buffer, ostr.str().c_str());
 }
 
+void project::GetFullBoxFileName(char * buffer, int buffer_size)
+{
+	ostringstream ostr;
+	
+	if (project_path != NULL) ostr << project_path << DIR_SEPARATOR;
+	ostr << project_filename << EXT_SEPARATOR << GetBoxFileNameExtension() << ends;
+	
+	if (strlen(ostr.str().c_str()) + 1 >= buffer_size)
+	{
+		assertion_failed(__FILE__, __LINE__, "buffer overflow!");
+	}
+	
+	strcpy(buffer, ostr.str().c_str());
+}
+
+void project::GetFullTrajFileName(char * buffer, int buffer_size)
+{
+	ostringstream ostr;
+	
+	if (project_path != NULL) ostr << project_path << DIR_SEPARATOR;
+	ostr << project_filename << EXT_SEPARATOR << GetTrajFileNameExtension() << ends;
+	
+	if (strlen(ostr.str().c_str()) + 1 >= buffer_size)
+	{
+		assertion_failed(__FILE__, __LINE__, "buffer overflow!");
+	}
+	
+	strcpy(buffer, ostr.str().c_str());
+}
+
+void project::GetFullPlotDataFileName(char * buffer, const char * plot_title, int buffer_size)
+{
+	ostringstream ostr;
+
+	if (project_path != NULL) ostr << project_path << DIR_SEPARATOR;
+	ostr << project_filename << "_" << plot_title << EXT_SEPARATOR << GetPlotDataFileNameExtension() << ends;
+	
+	if (strlen(ostr.str().c_str()) + 1 >= buffer_size)
+	{
+		assertion_failed(__FILE__, __LINE__, "buffer overflow!");
+	}
+	
+	strcpy(buffer, ostr.str().c_str());
+}
+
 /*##############################################*/
 /*##############################################*/
 
@@ -894,6 +959,17 @@
 	return true;
 }
 
+bool project::SavePlottingClient(base_wcl * wcl)
+{
+	p1dview_wcl* plot1d = dynamic_cast<p1dview_wcl *>(wcl);
+	if (plot1d)
+	{
+		new gtk_plot_data_save_dialog(this, plot1d);
+	}
+	
+	return true;
+}
+
 void project::UpdateAllViews(void)
 {
 	UpdateAllGraphicsViews();
@@ -4502,6 +4578,8 @@
 		
 	const char * s1 = "frame(num)"; const char * sv = "distance (nm)";
 	p1dview_wcl * plot = AddPlot1DClient(s1, sv, true);
+	std::string title = std::string("Coordinate Difference plot of ") + traj_filename;
+	plot->SetTitle(title.c_str());
 
 	float ekin;
 	float epot;
diff -Nur ./ghemical-3.0.0.original/src/project.h ./ghemical-3.0.0/src/project.h
--- ./ghemical-3.0.0.original/src/project.h	2023-07-04 19:36:42.509546452 +0000
+++ ./ghemical-3.0.0/src/project.h	2023-07-05 04:10:45.128736903 +0000
@@ -60,6 +60,8 @@
 using namespace std;
 
 #define FILENAME_FILTER	"*.gpr"
+#define PLOTFILENAME_FILTER	"*.txt"
+#define BOXFILENAME_FILTER	"*.box"
 
 /*################################################################################################*/
 
@@ -194,6 +196,9 @@
 	// ^^^^^^^^^^^^^^^^^^^^^^^^^
 	
 	const char * GetProjectFileNameExtension(void);
+	const char * GetBoxFileNameExtension(void);
+	const char * GetTrajFileNameExtension(void);
+	const char * GetPlotDataFileNameExtension(void);
 	
 	void SetProjectPath(const char *);
 	void SetProjectFileName(const char *);
@@ -205,6 +210,11 @@
 	void GetProjectFileName(char *, int, bool);
 	void GetFullProjectFileName(char *, int);
 
+	void GetFullBoxFileName(char *, int);
+	void GetFullTrajFileName(char *, int);
+	
+	void GetFullPlotDataFileName(char *, const char *, int);
+
 // methods for file I/O : ReadGPR and WriteGPR are friend functions so that it would be easier to "borrow" them elsewhere...
 // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 	friend bool ReadGPR_OLD(project &, istream &, bool, bool = false);		///< this is for the very old version.
@@ -298,6 +308,7 @@
 	rcpview_wcl * AddReactionCoordinatePlotClient(const char *, const char *, bool);
 	gpcview_wcl * AddGenericProteinChainClient(bool);
 	bool RemovePlottingClient(base_wcl *);
+	bool SavePlottingClient(base_wcl *);
 	
 	void UpdateAllViews(void);
 	
