diff -Nur ./libghemical-3.0.0.original/src/model.cpp ./libghemical-3.0.0/src/model.cpp
--- ./libghemical-3.0.0.original/src/model.cpp	2023-08-25 18:08:22.046854523 +0000
+++ ./libghemical-3.0.0/src/model.cpp	2023-08-25 18:21:47.758851449 +0000
@@ -5505,10 +5505,10 @@
 		{"gas_atom_went_down_out_from_membrane",""}, {"gas_atom_went_up_into_memrane",""}, {"gas_atom_went_down_into_membrane",""},
 		{"gas_atom_went_up_out_from_memrane",""}, {"coordinate_cumsum_velocity","nm"}, {"coordinate_cumsum_acceleration","nm"},
 		{"coordinate_cumsum_velocity_and_acceleration","nm"}, {"cumulative_mean_velocity","m/s"},
-		{"power", ""}
+		{"power", ""}, {"cumulative_mean_force",""}
 		};
 	
-	size_t crd_names_len = 23;
+	size_t crd_names_len = 24;
 
 	void * plot = fn_header(p, molgrouptype, ind_mol, dim, crd_type, crd_names, crd_names_len);
 
@@ -5980,6 +5980,9 @@
 			case 22:
 				value = power;
 				break;
+			case 23:
+				value = cumsum_force / (traj_frame_save_frq * (0 == loop ? 1 : loop));
+				break;
 
 			default:
 				break;

