diff -Nur ./libghemical-3.0.0/src/model.cpp ./libghemical-3.0.0.apply_bond/src/model.cpp
--- ./libghemical-3.0.0/src/model.cpp	2019-10-01 14:23:39.000000000 +0300
+++ ./libghemical-3.0.0.apply_bond/src/model.cpp	2019-10-01 13:59:36.000000000 +0300
@@ -51,6 +51,7 @@
 #include <cstring>
 #include <iomanip>
 #include <sstream>
+#include <strstream>
 using namespace std;
 
 #ifdef WIN32
@@ -290,6 +291,55 @@
 	exit(EXIT_FAILURE);
 }
 
+
+void model::SaveBox(const char * boxfilename)
+{
+	char mbuff1[256];
+	ofstream boxfile;
+	boxfile.open(boxfilename, ios::out);
+	ostrstream str3b(mbuff1, sizeof(mbuff1));
+	str3b << this->saved_periodic_box_HALFdim[0] << " ";
+	str3b << this->saved_periodic_box_HALFdim[1] << " ";
+	str3b << this->saved_periodic_box_HALFdim[2] << endl << ends;
+	boxfile << mbuff1;
+	boxfile.close();
+}
+
+bool model::LoadBox(const char * boxfilename)
+{
+	const int n = 1024;
+	char szBuff[n];
+	bool result = false;
+	FILE * stream;
+	stream = fopen(boxfilename, "rt");
+	if(stream)
+	{
+		if (!feof(stream))
+		{
+			char * ch;
+			ch = fgets(szBuff,n,stream);
+			if( ch != NULL && strlen(szBuff) > 1)
+			{
+				f64	box0, box1, box2;
+				if (3 == sscanf(szBuff, "%lf %lf %lf", 
+					&box0, 
+					&box1, 
+					&box2)) 
+				{
+					printf("box = [%f %f %f]\n", box0, box1, box2);
+
+					this->saved_periodic_box_HALFdim[0] = box0;
+					this->saved_periodic_box_HALFdim[1] = box1;
+					this->saved_periodic_box_HALFdim[2] = box2;
+					result = true;
+				}
+			}
+		}
+		fclose(stream);
+	}
+	return result;
+}
+
 void model::DiscardCurrEng(void)
 {
 //	cout << "discard!!!" << endl;
diff -Nur ./libghemical-3.0.0/src/model.h ./libghemical-3.0.0.apply_bond/src/model.h
--- ./libghemical-3.0.0/src/model.h	2019-10-01 14:23:39.000000000 +0300
+++ ./libghemical-3.0.0.apply_bond/src/model.h	2019-10-01 13:59:26.000000000 +0300
@@ -209,7 +209,10 @@
 	void DestroyRS(void);
 	
 	static void OpenLibDataFile(ifstream &, bool, const char *);
+
+    void SaveBox(const char * boxfilename);
+    bool LoadBox(const char * boxfilename);
 	
 // what to do for this one???
 // what to do for this one???
 // what to do for this one???
